src/eepromfs.c

```diff
@@ -0,0 +1,523 @@
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    /* Add an extra file for the "signature file" */
+    eepfs_files_count = count + 1;
+
+    /* Allocate the lookup table of file descriptors */
+    const size_t files_size = sizeof(eepfs_files[0]) * eepfs_files_count;
+    eepfs_files = (eepfs_file_t *)malloc(files_size);
+
+    if ( eepfs_files == NULL )
+    {
+        return EEPFS_ENOMEM;
+    }
+
+    /* The first file should always be the "signature file" */
+    const eepfs_file_t signature_file = { NULL, 0, 8 };
+    memcpy(&eepfs_files[0], &signature_file, sizeof(signature_file));
+
+    const char * file_path;
+    size_t file_size;
+    size_t total_blocks = 1;
+
+    /* Configure a file descriptor for each entry */
+    for ( size_t i = 1; i < eepfs_files_count; ++i )
+    {
+        file_path = entries[i - 1].path;
+        file_size = entries[i - 1].size;
+
+        /* Sanity check the file details */
+        if ( file_path == NULL || file_size == 0 )
+        {
+            eepfs_close();
+            return EEPFS_EBADINPUT;
+        }
+
+        /* Strip the leading '/' on paths for consistency */
+        file_path += ( file_path[0] == '/' );
+
+        /* Create a file descriptor and copy it into the table */
+        const eepfs_file_t entry_file = {
+            file_path,
+            total_blocks,
+            file_size,
+        };
+        memcpy(&eepfs_files[i], &entry_file, sizeof(entry_file));
+
+        /* Files must start on a block boundary */
+        total_blocks += divide_ceil(file_size, EEPROM_BLOCK_SIZE);
+    }
+
+    /* Ensure the filesystem will actually fit in available EEPROM */
+    if ( total_blocks > eeprom_total_blocks() )
+    {
+        eepfs_close();
+        return EEPFS_EBADFS;
+    }
+
+    /* Calculate and store the CRC-16 checksum for the declared entries */
+    const size_t entries_size = sizeof(eepfs_entry_t) * count;
+    eepfs_files_checksum = calculate_crc16((void *)entries, entries_size);
+
+    return EEPFS_ESUCCESS;
+}
+
+/**
+ * @brief De-initializes the EEPROM filesystem.
+ * 
+ * This cleans up the file lookup table.
+ * 
+ * You probably won't ever need to call this.
+ * 
+ * @return EEPFS_ESUCCESS on success or a negative error otherwise
+ */
+int eepfs_close(void)
+{
+    /* If eepfs was not initialized, don't do anything. */
+    if ( eepfs_files == NULL || eepfs_files_count == 0 )
+    {
+        return EEPFS_EBADFS;
+    }
+
+    /* Clear the file descriptor table */
+    free(eepfs_files);
+    eepfs_files = NULL;
+    eepfs_files_checksum = 0;
+    eepfs_files_count = 0;
+
+    return EEPFS_ESUCCESS;
+}
+
+/**
+ * @brief Reads an entire file from the EEPROM filesystem.
+ *
+ * @param[in]  path
+ *             Path of file in EEPROM filesystem to read from
+ * @param[out] dest
+ *             Buffer to read into
+ *
+ * @return EEPFS_ESUCCESS on success or a negative error otherwise
+ */
+int eepfs_read(const char * path, void * dest)
```

Comment:
I was wondering if `eepfs_read` (and `eepfs_write`) should receive a size argument, and the function should return error if the size is different from that expected by the filesystem.

It's obviously a redundant information but it's easy to provide by the caller (once you pass the buffer, you must have it allocated somehow of the correct size, so the size must be known at the point of the call) and can prevent stupid bugs and memory overruns which are always hard to debug on an embedded system.

Reply:
This is a good idea. I'll incorporate this.

Reply:
https://github.com/DragonMinded/libdragon/pull/125/commits/fd04a6359d93cd5f1d34906a97b287301ab69b55